docs/testing.md

```diff
@@ -197,4 +197,95 @@ val response1 = basicRequest.get(uri"http://api.internal/a").send()
 
 val response2 = basicRequest.post(uri"http://api.internal/b").send()
 // response2 will be whatever a "real" network call to api.internal/b returns
+```
+
+## Testing WebSockets
+
+Stub methods `whenRequestMatches` and `whenAnyRequest` also allow to specify behavior for `openWebsocket` call when used together with `thenRespondWebSocket` or `thenHandleOpenWebSocket`. `thenRespondWebSocket` is intended for cases when sttp supplied
+`WebSocketHandler` is used, that are cases where users code manipulates `WebSocket[F]` to implement logic.
+This way `thenRespondWebSocket` can be used to return test double `WebSocket`.
+
+Example:
+```scala
+val testWebSocket: WebSocket[Task] = ???
+
+implicit val testingBackend =
+  AsyncHttpClientZioBackend.stub
+    .whenAnyRequest
+    .thenRespondWebSocket(Headers(List.empty), testWebSocket)
+
+for {
+  handler <- ZioWebSocketHandler() // using sttp supplied handler
+  openResponse <- testingBackend.openWebsocket(basicRequest.get(uri"wss://some.uri"), handler)
+  webSocket = openResponse.result // interaction with WebSocket is interesting in this case
+  message <- webSocket.receive
+} yield message
+```
+
+## WebSocketStub
+
+Implementation of `WebSocket` stub is available. It wraps state machine that can be used to simulate simple WebSockets interactions.
+It allows to set initial responses for `receive` calls as well as logic to add further messages in reaction to `send` calls.
+`SttpBackendStub` has special API that accepts `WebSocketStub` and builds it to `WebSocket` instance.
+
+Example:
+```
+val backend = SttpBackendStub.synchronous[Identity]
+val webSocketStub = WebSocketStub
+  .withInitialIncoming(
+    List(WebSocketFrame.text("Hello from the server!"))
+  )
+  .thenRespondS(0) {
+    case (counter, tf: WebSocketFrame.Text) => (counter + 1, List(WebSocketFrame.text(s"echo: ${tf.payload}")))
+    case (counter, _)                       => (counter, List.empty)
+  }
+backend.whenAnyRequest.thenRespondWebSocket(Headers(List.empty), webSocketStub)
+```
+There is a possiblity to add error responses as well, if it's not enough,
```

Comment:
There's a lot of extra line breaks in the middle of the sentences, throughout this file - they should be removed.